File: src/frontend/folha/lib/app/data/model/folha_plano_saude_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:folha/app/infra/infra_imports.dart';
import 'package:folha/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:folha/app/data/domain/domain_imports.dart';

class FolhaPlanoSaudeModel {
	int? id;
	int? idOperadoraPlanoSaude;
	int? idColaborador;
	DateTime? dataInicio;
	String? beneficiario;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;
	OperadoraPlanoSaudeModel? operadoraPlanoSaudeModel;

	FolhaPlanoSaudeModel({
		this.id,
		this.idOperadoraPlanoSaude,
		this.idColaborador,
		this.dataInicio,
		this.beneficiario,
		this.viewPessoaColaboradorModel,
		this.operadoraPlanoSaudeModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'data_inicio',
		'beneficiario',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Data Inicio',
		'Beneficiario',
	];

	FolhaPlanoSaudeModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idOperadoraPlanoSaude = jsonData['idOperadoraPlanoSaude'];
		idColaborador = jsonData['idColaborador'];
		dataInicio = jsonData['dataInicio'] != null ? DateTime.tryParse(jsonData['dataInicio']) : null;
		beneficiario = FolhaPlanoSaudeDomain.getBeneficiario(jsonData['beneficiario']);
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
		operadoraPlanoSaudeModel = jsonData['operadoraPlanoSaudeModel'] == null ? OperadoraPlanoSaudeModel() : OperadoraPlanoSaudeModel.fromJson(jsonData['operadoraPlanoSaudeModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idOperadoraPlanoSaude'] = idOperadoraPlanoSaude != 0 ? idOperadoraPlanoSaude : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['dataInicio'] = dataInicio != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataInicio!) : null;
		jsonData['beneficiario'] = FolhaPlanoSaudeDomain.setBeneficiario(beneficiario);
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
		jsonData['operadoraPlanoSaudeModel'] = operadoraPlanoSaudeModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idOperadoraPlanoSaude = plutoRow.cells['idOperadoraPlanoSaude']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		dataInicio = Util.stringToDate(plutoRow.cells['dataInicio']?.value);
		beneficiario = plutoRow.cells['beneficiario']?.value != '' ? plutoRow.cells['beneficiario']?.value : '1=Somente Colaborador';
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
		operadoraPlanoSaudeModel = OperadoraPlanoSaudeModel();
		operadoraPlanoSaudeModel?.nome = plutoRow.cells['operadoraPlanoSaudeModel']?.value;
	}	

	FolhaPlanoSaudeModel clone() {
		return FolhaPlanoSaudeModel(
			id: id,
			idOperadoraPlanoSaude: idOperadoraPlanoSaude,
			idColaborador: idColaborador,
			dataInicio: dataInicio,
			beneficiario: beneficiario,
		);			
	}

	
}

File: backend/folha/src/model/folha_plano_saude_model.py
from src import db
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel
from src.model.operadora_plano_saude_model import OperadoraPlanoSaudeModel


class FolhaPlanoSaudeModel(db.Model):
    __tablename__ = 'folha_plano_saude'

    id = db.Column(db.Integer, primary_key=True)
    data_inicio = db.Column(db.DateTime)
    beneficiario = db.Column(db.String(1))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))
    id_operadora_plano_saude = db.Column(db.Integer, db.ForeignKey('operadora_plano_saude.id'))

    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])
    operadora_plano_saude_model = db.relationship('OperadoraPlanoSaudeModel', foreign_keys=[id_operadora_plano_saude])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_operadora_plano_saude = data.get('idOperadoraPlanoSaude')
        self.id_colaborador = data.get('idColaborador')
        self.data_inicio = data.get('dataInicio')
        self.beneficiario = data.get('beneficiario')

    def serialize(self):
        return {
            'id': self.id,
            'idOperadoraPlanoSaude': self.id_operadora_plano_saude,
            'idColaborador': self.id_colaborador,
            'dataInicio': self.data_inicio.isoformat(),
            'beneficiario': self.beneficiario,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
            'operadoraPlanoSaudeModel': self.operadora_plano_saude_model.serialize() if self.operadora_plano_saude_model else None,
        }